tokio/src/task/consume_budget.rs

```diff
@@ -0,0 +1,33 @@
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// Consumes a unit of budget and returns the execution back to the Tokio
+/// runtime *if* the task went out ouf budget.
+///
+/// The task will only yield if it ran out of its coop budget.
+/// It can be used in order to insert optional yield points into long
+/// computations that do not use Tokio resources like sockets or semaphores,
+/// without redundantly yielding to runtime each time.
+#[cfg_attr(docsrs, doc(cfg(feature = "rt")))]
+pub async fn consume_budget() {
+    struct ConsumeBudget {
+        status: Poll<()>,
+    }
+
+    impl Future for ConsumeBudget {
+        type Output = ();
+
+        fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
+            if self.status.is_ready() {
+                return self.status;
+            }
+            self.status = crate::coop::poll_proceed(cx).map(|restore| {
+                restore.made_progress();
+            });
+            self.status
+        }
+    }
+
+    ConsumeBudget { status: Poll::Pending }.await
```

Comment:
nit, take it or leave it: i think we could _probably_ implement this a bit more simply using [`poll_fn`](https://github.com/tokio-rs/tokio/blob/ac0f894dd9be2a3a7e86405607ae240e9bc0a770/tokio/src/future/poll_fn.rs):
```suggestion
    let mut status = Poll::Pending;
    
    crate::future::poll_fn(move |cx| {
        if status.is_ready() {
            return status;
        }

        status = crate::coop::poll_proceed(cx).map(|restore| {
            restore.made_progress();
        });

        status
    }).await
```
i believe this should work identically, and avoids a little bit of boilerplate by not having to `impl Future` for a type.

Reply:
Neat! Looks much better, thanks